### ISSUE TITLE
Add support for sk- (hardware-backed) ed25519 keys

### ISSUE THREAD

**Issue by anoadragon453** [Mic92/ssh-to-age#35]
OpenSSH 8.2 or higher [supports](https://undeadly.org/cgi?action=article;sid=20191115064850) hardware-backed variants of ed25519 and ecdsa SSH key types, otherwise known as "secure key" (sk) variants.

I have a Yubikey and use its [FIDO2 support](https://developers.yubico.com/SSH/Securing_SSH_with_FIDO2.html) to store a `sk-ed25519` SSH key on the device. It is not (supposedly) possible to get access to the private keys of this variant, but you can access the public key of course.

I use this key to authenticate to my servers, and I'd like to use it with sops-nix as well. However, it appears that `ssh-to-age` does not support the `-sk` variant of ed25519 keys, so I get an error:

```
$ ssh-to-age -i ~/.ssh/id_ed25519_sk_rk.pub
ssh-to-age: failed to convert 'sk-ssh-ed25519@openssh.com AAAAGnNrLXNzaC1lZDI1NTE5QG9wZW5zc2guY29tAAAAIGVdcgCRUwCd83w5L+k5yhDHrLDF88GgDWdhvMqYAUiAAAAABHNzaDo= ssh:': failed to parse ssh public key: illegal base64 data at input byte 3
```

which I believe is due to the check for the `"ssh-"` prefix here:

https://github.com/Mic92/ssh-to-age/blob/3d775f6681c7b9d001200516ff20c5e5a3e94b58/convert.go#L99-L103

If we expand it to:

```go
	if strings.HasPrefix(string(sshKey), "ssh-") || strings.HasPrefix(string(sshKey), "sk-ssh-") {
		pk, _, _, _, err = ssh.ParseAuthorizedKey(sshKey)
	} else {
		_, _, pk, _, _, err = ssh.ParseKnownHosts(sshKey)
	}
```

then we predictably get:

```
$ ssh-to-age -i ~/.ssh/id_ed25519_sk_rk.pub  
skipped key: got sk-ssh-ed25519@openssh.com key type, but only ed25519 keys are supported
```

due to:

https://github.com/Mic92/ssh-to-age/blob/3d775f6681c7b9d001200516ff20c5e5a3e94b58/convert.go#L107-L110

if we then expand _that_ code to:

```go
	// We only care about ed25519 (and the secure key variant)
	if pk.Type() != ssh.KeyAlgoED25519 && pk.Type() != ssh.KeyAlgoSKED25519 {
		return nil, fmt.Errorf("got %s key type, but %w", pk.Type(), UnsupportedKeyType)
	}
```

we then get:

```
$ ssh-to-age -i ~/.ssh/id_ed25519_sk_rk.pub
/nix/store/lngv3dl3yal64xripysv284781qrpxj9-ssh-to-age-1.1.1/bin/ssh-to-age: failed to convert 'sk-ssh-ed25519@openssh.com AAAAGnNrLXNzaC1lZDI1NTE5QG9wZW5zc2guY29tAAAAIGVdcgCRUwCd83w5L+k5yhDHrLDF88GgDWdhvMqYAUiAAAAABHNzaDo= ssh:': BUG! public key does not implement ssh.CryptoPublicKey
```

which makes sense. [`ssh.skEd25519PublicKey`](https://github.com/golang/crypto/blob/2e82bdd1719d8f50d18bbb60acc52accc71330b1/ssh/keys.go#L820-L825) does not implement `ssh.CryptoPublicKey`, while [`ssh.ed25519PublicKey` does](https://github.com/golang/crypto/blob/2e82bdd1719d8f50d18bbb60acc52accc71330b1/ssh/keys.go#L608-L610).

I'm wondering if it's still possible to get the elliptic curve points out of `pk` and convert it to an age key regardless? Or am I barking up the wrong tree and should go pester to go devs to implement `ssh.CryptoPublicKey` on `ssh.skEd25519PublicKey`? :)

**Comment by Mic92** [comment]
The bigger issue here is that sops also needs to support the yubikey for encryption, which is currently not the case: https://github.com/getsops/sops/issues/1103
So yes we could add the conversion here but we would still need support in sops's age implementation to make actually use of it.

**Comment by Mic92** [comment]
Sops cannot decrypt because it would have to reach out for the yubikey, which is not implemented.

**Comment by anoadragon453** [comment]
Right, that part makes sense to me. Thanks for responding.

I'll wait on that front first then. Feel free to leave this issue open in the meantime or close it if you like.

**Comment by Mic92** [comment]
I'll keep it open because otherwise someone else will just come again with the same issue.

**Comment by haslersn** [comment]
AFAIK, what this issue asks for is impossible: Even if ssh-to-age would support converting a hardware-backed key, the Yubikey won't let you convert the private key and won't let you do the operation required for decryption with that key.

ssh-to-age converts an ed25519 key (usable for signing) to an x25519 key (usable for encryption). Both the public and private key need to be converted. The Yubikey won't let you do this conversion of the private key and then it obviously also won't let you do the Diffie-Hellman key exchange required for decryption.

Correct me if I'm wrong.

**Comment by gauravjuvekar** [comment]
Yes haslersn is correct. sk- variants are fido based and the security key will only generate signatures from the token's private key. It will not decrypt anything (or allow the private key to leave the hardware token obviously).

You will have to use PGP / PIV based hardware tokens (some of which support generating and sealing the private key entirely on the hardware token resulting in a similar security posture) for decryption.

**Comment by Mic92** [comment]
Sops now supports age plugins. So instead of sk-ed25519 keys I would suggest using https://github.com/olastor/age-plugin-fido2-hmac

**Comment by gauravjuvekar** [comment]
Note, in fido2-hmac, the private key of the recipient leaves the hardware token.

**Comment by Mic92** [comment]
> Note, in fido2-hmac, the private key of the recipient leaves the hardware token.

Which doesn't really makes a big difference because the sops private key is symmetric and stored in the document.
